File: solution/java/day13/src/test/java/blog/ArticleBuilder.java
package blog;

import java.util.HashMap;
import java.util.Map;

public class ArticleBuilder {
    public static final String AUTHOR = "Pablo Escobar";
    public static final String COMMENT_TEXT = "Amazing article !!!";
    private final HashMap<String, String> comments;

    public ArticleBuilder() {
        comments = new HashMap<>();
    }

    public static ArticleBuilder anArticle() {
        return new ArticleBuilder();
    }

    public ArticleBuilder commented() {
        this.comments.put(COMMENT_TEXT, AUTHOR);
        return this;
    }

    public Article build() throws CommentAlreadyExistException {
        var article = new Article(
                "Lorem Ipsum",
                "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );
        for (Map.Entry<String, String> comment : comments.entrySet()) {
            article.addComment(comment.getKey(), comment.getValue());
        }
        return article;
    }
}

File: solution/kotlin/day19/src/test/kotlin/ArticleBuilder.kt
import blog.Article

class ArticleBuilder {
    private val comments: MutableMap<String, String> = mutableMapOf()

    fun commented(): ArticleBuilder = apply { comments += (COMMENT_TEXT to AUTHOR) }

    fun build(): Article =
        comments.entries
            .fold(
                Article("Lorem Ipsum", "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore")
            ) { article, comment -> article.addComment(comment.key, comment.value).getOrNull()!! }

    companion object {
        const val AUTHOR = "Pablo Escobar"
        const val COMMENT_TEXT = "Amazing article !!!"
        fun anArticle(): ArticleBuilder = ArticleBuilder()
    }
}

